email-service/internal/broker/transport/consumer.go

```diff
@@ -0,0 +1,122 @@
+package transport
+
+import (
+	"fmt"
+	"log/slog"
+	"sync"
+
+	"github.com/GenesisEducationKyiv/software-engineering-school-4-0-Hukyl/email-service/internal/broker/transport/config"
+	amqp "github.com/rabbitmq/amqp091-go"
+)
+
+func logAndWrap(msg string, err error) error {
+	slog.Error(msg, slog.Any("error", err))
+	return fmt.Errorf("%s: %w", msg, err)
+}
+
+var listenerAccess = sync.Mutex{}
+
+type Listener func([]byte) error
+
+type Consumer struct {
```

Comment:
Do you send acknowledgments?

Reply:
Auto acknowledgment is set when initializing new Consumer (in amqp.Channel.Consume), but probably it is best to handle an error from each mailer and do not send acknowledgments if the message has failed (i.e. retry)